### DISCUSSION TITLE
Use postgres with useFakeTimer

### DISCUSSION THREAD

**Discussion by mehdi-valette** [porsager/postgres#798] [Q&A] [answered]
Hello everyone! 

At my job we've been using postgres for a few months and it works great, congrats 👍

Recently we hit a wall: in a test (vitest) we use `useFakeTimers`, which mocks the javascript date and time objects (https://vitest.dev/api/vi.html#vi-usefaketimers)
When we call any database query (transaction, insert, query, ...) the script is blocked and the test times out.

Does anyone have an idea why that's the case and how to fix it?

**Comment by joaomarcosjova** [comment]
To address the issue of `useFakeTimers` blocking your PostgreSQL queries in your test environment, we'll need to take several steps. Below, I'll outline the code for setting up and troubleshooting your test environment:

### 1. Set Up Test Environment

First, ensure you have the necessary dependencies installed:

```bash
npm install --save-dev vitest pg
```

### 2. Initialize PostgreSQL Connection

In your test file, establish a connection to your PostgreSQL database using the `pg` library:

```javascript
const { Pool } = require('pg');

const pool = new Pool({
  user: 'your_username',
  host: 'localhost',
  database: 'your_database',
  password: 'your_password',
  port: 5432, // Default PostgreSQL port
});
```

### 3. Write Test Cases

Write your test cases using `useFakeTimers` and execute PostgreSQL queries within them:

```javascript
const { useFakeTimers } = require('vitest');

describe('PostgreSQL Tests', () => {
  let clock;

  before(() => {
    // Initialize fake timers before tests
    clock = useFakeTimers();
  });

  after(() => {
    // Restore original timers after tests
    clock.restore();
  });

  it('should perform a database query', async () => {
    // Mock PostgreSQL query
    const res = await pool.query('SELECT NOW()');
    console.log('Current time from PostgreSQL:', res.rows[0].now);
  });
});
```

### 4. Troubleshoot Blocking Issue

If the tests are timing out due to `useFakeTimers` blocking PostgreSQL queries, consider the following troubleshooting steps:

- Ensure that your PostgreSQL database is properly configured and accessible from your test environment.
- Check for any long-running queries or inefficient database operations that might be causing the blocking issue.
- Review the documentation and source code for `useFakeTimers` to understand its behavior and limitations.
- Experiment with alternative testing strategies, such as mocking database queries or using a different test framework.

### Conclusion

By following these steps and troubleshooting techniques, you should be able to identify and resolve the issue of `useFakeTimers` blocking your PostgreSQL queries in your test environment. Remember to test your changes thoroughly and monitor the test results for any improvements or regressions.

**Reply by mehdi-valette** [reply]
Hi Joao, thank you for your fast response!

When I use `new pool` from 'pg' it works, that's a start 👍

However in the code we assign the client with `postgres(CONNECTION_STRING, options)`.

In the code below, we never reach the line "debugger". However we can reach it if we comment `vi.useFakerTimers()` and `vi.setSystemTime(new Date())`.

```typescript
import postgres from 'postgres';
import { afterAll, beforeAll, expect, test, vi } from 'vitest';

const pgClient = postgres(process.env.CONNECTION_STRING, {});

beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date());
});

afterAll(() => {
    vi.useRealTimers();
});


test('lock', async () => {
    const result2 = await pgClient`SELECT * FROM "user"`;

  debugger; // never reaches this line
});
```

Do you know why that would be the case?

**Reply by joaomarcosjova** [reply]
In the provided code, the issue seems to be related to how the vi module from vitest is manipulating time using fake timers. When you use vi.useFakeTimers() and vi.setSystemTime(new Date()), the timing of asynchronous operations, such as database queries, might be affected.

Here are a few potential reasons why the debugger statement is not being reached:

1. Timing conflict: When using fake timers, asynchronous operations might not resolve within the expected time frame, causing the test to hang or not reach the debugger statement.

2. Synchronous execution: It's possible that the debugger statement is being skipped due to the synchronous nature of the code execution. The debugger statement might not be reached if there are no asynchronous operations or if they complete very quickly.

3. Error handling: If there's an error in the pgClient query, it might be caught by the test framework or handled internally, preventing the debugger statement from being reached.

To troubleshoot the issue, you can try the following steps:

1. Check for errors: Ensure that there are no errors in the pgClient query or any other part of the test code that might prevent the debugger statement from being reached.

2. Increase timeout: If the test framework has a timeout setting, try increasing it to allow more time for asynchronous operations to complete.

3. Debug asynchronously: Instead of using a debugger statement, try adding console.log statements or using the debugger tool in your IDE to debug the asynchronous code flow.

4. Simplify the test: Temporarily remove the use of fake timers and other complex setup to see if the debugger statement is reached in a simpler test environment.

By debugging and simplifying your test setup, you can identify the root cause of the issue and ensure that the debugger statement is reached as expected.

```javascript
import postgres from 'postgres';
import { afterAll, beforeAll, test, vi } from 'vitest';

const pgClient = postgres('postgres://wedo:test@localhost:5432/wedo_test', {});

beforeAll(() => {
    // Set up fake timers before all tests
    vi.useFakeTimers();
    vi.setSystemTime(new Date());
});

afterAll(() => {
    // Reset to real timers after all tests
    vi.useRealTimers();
});

test('lock', async () => {
    // Make a database query using pgClient
    const result2 = await pgClient`SELECT * FROM "user"`;

    // Delay the execution of debugger to ensure the database query completes
    setTimeout(() => {
        debugger; // Reach this line after database query completion
    }, 1000); // Adjust the delay as needed
});
```

In this version:

- We're using `setTimeout` to delay the execution of the `debugger` statement. This delay allows the database query to complete before reaching the `debugger` line.
- Adjust the timeout value (`1000` milliseconds in this example) based on the expected execution time of the database query. It's essential to ensure that the delay is sufficient for the asynchronous operation to finish.
- Use logging statements or console outputs to track the flow of execution and ensure that the database query completes successfully before the `debugger` statement is reached.

This approach helps synchronize the timing of asynchronous operations with the manipulation of fake timers, ensuring that your test code behaves predictably and reliably.

**Reply by mehdi-valette** [reply]
``Hello, thank you for your answer.

Here's a review:

# potential reasons

> Timing conflict: When using fake timers, asynchronous operations might not resolve within the expected time frame, causing the test to hang or not reach the debugger statement.

That's what's happening I think, but why is that the case and how can we solve it?

> Synchronous execution: It's possible that the debugger statement is being skipped due to the synchronous nature of the code execution. The debugger statement might not be reached if there are no asynchronous operations or if they complete very quickly.

The code execution should reach the `debugger` statement because it's an `async` function and there's an `await` before the call to the asynchronous code. Plus the `debugger` statement is reached when we don't use `useFaketimers`

> Error handling: If there's an error in the pgClient query, it might be caught by the test framework or handled internally, preventing the debugger statement from being reached.

The query is correct because it executes correctly when we don't use `useFakeTimers`. But just in case I made a new test case to make sure of that.


# Troubleshooting
> Check for errors: Ensure that there are no errors in the pgClient query or any other part of the test code that might prevent the debugger statement from being reached.

If I remove `useFaketimer` everything works, so the problem is only with `useFakeTimers`.

> Increase timeout: If the test framework has a timeout setting, try increasing it to allow more time for asynchronous operations to complete.

I increased it to 20s, which should be more than enough for such a small query on such a small database (all tables are empty at the beginning of the test; which should not cause any problem).

> Debug asynchronously: Instead of using a debugger statement, try adding console.log statements or using the debugger tool in your IDE to debug the asynchronous code flow.

I added `console.log` statements, the result is identical (see below).

> Simplify the test: Temporarily remove the use of fake timers and other complex setup to see if the debugger statement is reached in a simpler test environment.

The test is as simple as it can be, and it works if I remove `useFaketimers`.

# Code and result using useFakeTimers
```typescript
import postgres from 'postgres';
import { afterAll, beforeAll, test, vi } from 'vitest';

const pgClient = postgres(process.env.PG_CONNECTION_STRING, {});

beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date());
});

afterAll(() => {
    vi.useRealTimers();
});

test('lock', async () => {
    try {
        console.log('before call');
        const result2 = await pgClient`SELECT * FROM "user"`;
        console.log('after call');
    } catch (e) {
        console.log('catch');
    }

    console.log('end');
}, 20_000);
```

result:
```
before call

Error: Test timed out in 20000ms.
```


# Code and result without useFakeTimers
```typescript
import postgres from 'postgres';
import { afterAll, beforeAll, test, vi } from 'vitest';

const pgClient = postgres(process.env.PG_CONNECTION_STRING, {});

beforeAll(() => {
    // vi.useFakeTimers();  ------------------ COMMENTED  useFakeTimers ----------------
    vi.setSystemTime(new Date());
});

afterAll(() => {
    vi.useRealTimers();
});

test('lock', async () => {
    try {
        console.log('before call');
        const result2 = await pgClient`SELECT * FROM "user"`;
        console.log('after call');
    } catch (e) {
        console.log('catch');
    }

    console.log('end');
}, 20_000);
```

Result:
```
before call
after call
end
```

**Comment by mehdi-valette** [comment, marked as answer]
For the record, it works with this code:

```typescript
import postgres from 'postgres';
import { afterAll, beforeAll, test, vi } from 'vitest';

const pgClient = postgres(process.env.PG_CONNECTION_STRING, {});

beforeAll(() => {
    vi.useFakeTimers({toFake: ['Date']}); // mock only what you need
    vi.setSystemTime(new Date());
});

afterAll(() => {
    vi.useRealTimers();
});

test('lock', async () => {
    try {
        console.log('before call');
        const result2 = await pgClient`SELECT * FROM "user"`;
        console.log('after call');
    } catch (e) {
        console.log('catch');
    }

    console.log('end');
}, 20_000);
```

I suppose useFakeTimers mocks something that is used by postgres in a way that disturbs postgres, thus creating an issue.
